### PR TITLE
Fix VQD callback not being used

### LINKED ISSUE #8575: Callback function in VQD is defined but never used
### What should we add?

The Variational Quantum Deflation algorithm has an attribute `self._callback` which is never used in `vqd.solve()`.
Inspired by the code for the VQE algorithm, the callback should be called in the `energy_evaluation()` call.
It also seems interesting that the callback also informs the user of the current step (VQD runs k steps of VQE).

Currently, the test `test_callback` checks the equality of empty lists
```
history = {'eval_count': [], 'parameters': [], 'mean': [], 'std': []}
```

The change would consist in adding the following lines during the call of the `energy_evaluation()`. These lines were taken from the VQE algorithm.
```
if self._callback is not None:
    variance = np.real(expectation.compute_variance(sampled_expect_op))
    estimator_error = np.sqrt(variance / self.quantum_instance.run_config.shots)
    for i, param_set in enumerate(parameter_sets):
        self._eval_count += 1
        self._callback(self._eval_count, param_set, means[i], estimator_error[i], step)
else:
    self._eval_count += len(means)
```
Giving
```
history {
'eval_count': [1, 2, 3, 1, 2, 3],
'parameters': [array([ 3.61186007,  4.19301252,  0.6019852 ,  5.94953681, -3.30704704,
        1.84629318, -5.4660436 ,  0.6984088 ]), array([ 4.61186007,  4.19301252,  0.6019852 ,  5.94953681, -3.30704704,
        1.84629318, -5.4660436 ,  0.6984088 ]), array([ 4.61186007,  5.19301252,  0.6019852 ,  5.94953681, -3.30704704,
        1.84629318, -5.4660436 ,  0.6984088 ]), array([ 3.61186007,  4.19301252,  0.6019852 ,  5.94953681, -3.30704704,
        1.84629318, -5.4660436 ,  0.6984088 ]), array([ 4.61186007,  4.19301252,  0.6019852 ,  5.94953681, -3.30704704,
        1.84629318, -5.4660436 ,  0.6984088 ]), array([ 3.61186007,  5.19301252,  0.6019852 ,  5.94953681, -3.30704704,
        1.84629318, -5.4660436 ,  0.6984088 ])], 
'mean': [-1.0632369218371718, -1.4565924844031897, -1.360034418572368, 37.340083390662826,        48.54340751559681, 28.58589345125128], 
'std': [0.011432088966174363, 0.009556915198516208, 0.013956731222284432, 0.011432088966174363, 0.009556915198516208, 0.015443719747907501], 
'step': [1, 1, 1, 2, 2, 2]
}

```

### PR DESCRIPTION
**Pull request by Anthony-Gandon**
Fix #8575 
This small change consists in adding the following lines during the call of the energy_evaluation() in VQD. These lines were taken from the VQE algorithm.
```
if self._callback is not None:
    variance = np.real(expectation.compute_variance(sampled_expect_op))
    estimator_error = np.sqrt(variance / self.quantum_instance.run_config.shots)
    for i, param_set in enumerate(parameter_sets):
        self._eval_count += 1
        self._callback(self._eval_count, param_set, means[i], estimator_error[i], step)
else:
    self._eval_count += len(means)
```
The corresponding test is modified accordingly.
The callback also keeps track of the current step in the VQD algorithm.

### CODE REVIEW

**Review by jlapeyre — CHANGES_REQUESTED** [review]
LGTM.  I did have a small question.

File: qiskit/algorithms/eigen_solvers/vqd.py [hunk 1]
```diff
@@ -108,10 +108,11 @@ def __init__(
         Args:
             ansatz: A parameterized circuit used as ansatz for the wave function.
             k: the number of eigenvalues to return. Returns the lowest k eigenvalues.
-            betas: beta parameter in the VQD paper. Should have size k -1, the number of excited states.
-                It is a hyperparameter that balances the contribution of the overlap
-                term to the cost function and has a default value computed as
-                mean square sum of coefficients of observable.
+            betas: beta parameters in the VQD paper.
+                Should have size k -1, the number of excited states.
```
[the remark this answers is not in the document]
Comment (Anthony-Gandon) [reply qiskit/algorithms/eigen_solvers/vqd.py]: Indeed. I changed it accordingly

**Review by Cryoris — APPROVED** [review]
LGTM thanks for the fix!

**Review by woodsp-ibm — APPROVED** [review]

**Review by jlapeyre — APPROVED** [review]
Thanks!

**Pull request merged**